core/domain/exp_domain.py

```diff
@@ -474,6 +474,273 @@ def __init__(self, change_list):
+            exp_version_diffs: list(ExplorationVersionsDiff). The specific
+                changes made to the exploration at each corresponding version.
+            state_equality_predicate:
+                callable(state_domain.State, state_domain.State) -> bool | None.
+                Returns True when two states are "equal". The predicate is used
+                to enforce the following invariant: all versions of the state in
+                a mapping are equivalent. If None, then uses a predicate that
+                unconditionally returns True.
+        """
+        if not exps or not exp_version_diffs:
+            raise ValueError('Inputs must be non-empty')
+
+        if len(exps) != len(exp_version_diffs):
+            raise ValueError('len of inputs do not match')
+
+        actual_versions = [exp.version for exp in exps]
+        expected_versions = list(python_utils.RANGE(1, len(exps) + 1))
+        if actual_versions != expected_versions:
+            raise ValueError(
+                'Explorations must be sorted, begin from version 1, and have '
+                'no gaps; but got invalid input versions == [%s]' % (
+                    ', '.join(
+                        python_utils.UNICODE(v) for v in actual_versions)))
+
+        self._state_spans = []
+        for exp, diff in python_utils.ZIP(exps, exp_version_diffs):
+            new_state_spans = dict()
+            for state_name, state_content in exp.states.items():
+                if exp.version == 1 or state_name in diff.added_state_names:
+                    new_state_spans[state_name] = StateVersionSpan(
+                        exp.version, state_name, state_content,
+                        state_equality_predicate=state_equality_predicate)
+                elif state_name not in diff.deleted_state_names:
+                    prev_state_spans, prev_state_name = (
+                        self._state_spans[-1],
+                        diff.new_to_old_state_names.get(state_name, state_name))
+                    new_state_spans[state_name] = (
+                        prev_state_spans[prev_state_name].extend_or_split(
+                            exp.version, state_name, state_content, diff))
+            self._state_spans.append(new_state_spans)
+
+    def get_state_span(self, exp_version, state_name):
+        """Returns the span of versions for the given state which includes
+        the given exploration version.
+
+        Args:
+            exp_version: int. The exploration version to get.
+            state_name: str. The name of the state at the given exploration
+                version.
+
+        Returns:
+            StateVersionSpan | None. The span of versions for the given
+            state which includes the given exploration version. Returns None if
+            the state did not exist at that version.
+        """
+        return self._state_spans[exp_version - 1].get(state_name, None)
```

Comment:
Maybe add a comment explaining the need for the "-1". Ditto below.

Or add a private method along the lines of self._get_state_span(version)